packages/eslint-plugin/src/rules/restrict-template-expressions.ts

```diff
@@ -12,16 +12,35 @@ import {
   isTypeNeverType,
 } from '../util';
 
-type Options = [
-  {
-    allowAny?: boolean;
-    allowBoolean?: boolean;
-    allowNullish?: boolean;
-    allowNumber?: boolean;
-    allowRegExp?: boolean;
-    allowNever?: boolean;
+const optionsObj = {
+  Any: isTypeAnyType,
+  Array: (type, checker, rec): boolean => {
+    if (!checker.isArrayType(type)) {
+      return false;
+    }
+    const numberIndexType = type.getNumberIndexType();
```

Comment:
[Style] Nit: a little more terse...

```suggestion
    const numberIndexType = checker.isArrayType(type) && type.getNumberIndexType();
```

Reply:
nice! not sure why I didn't think of this haha